File: src/main/java/me/creepinson/creepinoutils/util/client/gui/Checkbox.java
/*
 * Copyright (c) Creepinson
 */

package me.creepinson.creepinoutils.util.client.gui;

import com.mojang.blaze3d.platform.GlStateManager;
import com.mojang.blaze3d.systems.RenderSystem;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.FontRenderer;
import net.minecraft.client.gui.widget.button.Button;
import net.minecraft.util.ResourceLocation;
import net.minecraft.util.math.MathHelper;
import net.minecraftforge.api.distmarker.Dist;
import net.minecraftforge.api.distmarker.OnlyIn;

/**
 * @author Creepinson
 */
@OnlyIn(Dist.CLIENT)
public class Checkbox extends Button {
    private static final ResourceLocation TEXTURE = new ResourceLocation("textures/gui/checkbox.png");
    boolean checked;

    public Checkbox(int xIn, int yIn, int widthIn, int heightIn, String msg, boolean isChecked, Button.IPressable onPress) {
        super(xIn, yIn, widthIn, heightIn, msg, onPress);
        this.checked = isChecked;
    }

    @Override
    public void onPress() {
        this.checked = !this.checked;
        super.onPress();
    }

    public boolean isChecked() {
        return this.checked;
    }

    @Override
    public void renderButton(int p_renderButton_1_, int p_renderButton_2_, float p_renderButton_3_) {
        Minecraft minecraft = Minecraft.getInstance();
        minecraft.getTextureManager().bindTexture(TEXTURE);
        RenderSystem.enableDepthTest();
        FontRenderer fontrenderer = minecraft.fontRenderer;
        RenderSystem.color4f(1.0F, 1.0F, 1.0F, this.alpha);
        RenderSystem.enableBlend();
        RenderSystem.defaultBlendFunc();
        RenderSystem.blendFunc(GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA);
        blit(this.x, this.y, 0.0F, this.checked ? 20.0F : 0.0F, 20, this.height, 32, 64);
        this.renderBg(minecraft, p_renderButton_1_, p_renderButton_2_);
        int i = 14737632;
        this.drawString(fontrenderer, this.getMessage(), this.x + 24, this.y + (this.height - 8) / 2, 14737632 | MathHelper.ceil(this.alpha * 255.0F) << 24);
    }
}

File: src/main/kotlin/com/theoparis/creepinoutils/util/client/gui/Checkbox.kt
/*
 * Copyright (c) Creepinson
 */
package com.theoparis.creepinoutils.util.client.gui

import com.mojang.blaze3d.matrix.MatrixStack
import com.mojang.blaze3d.platform.GlStateManager
import com.mojang.blaze3d.systems.RenderSystem
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.widget.button.Button
import net.minecraft.util.ResourceLocation
import net.minecraft.util.math.MathHelper
import net.minecraft.util.text.ITextComponent
import net.minecraftforge.api.distmarker.Dist
import net.minecraftforge.api.distmarker.OnlyIn

/**
 * @author Creepinson
 */
@OnlyIn(Dist.CLIENT)
class Checkbox(
    xIn: Int,
    yIn: Int,
    widthIn: Int,
    heightIn: Int,
    msg: ITextComponent?,
    var isChecked: Boolean,
    onPress: IPressable?
) : Button(xIn, yIn, widthIn, heightIn, msg, onPress) {
    override fun onPress() {
        isChecked = !isChecked
        super.onPress()
    }

    override fun renderButton(matrixStack: MatrixStack, mouseX: Int, mouseY: Int, partialTicks: Float) {
        val minecraft = Minecraft.getInstance()
        minecraft.getTextureManager().bindTexture(TEXTURE)
        RenderSystem.enableDepthTest()
        val fontrenderer = minecraft.fontRenderer
        RenderSystem.color4f(1.0f, 1.0f, 1.0f, alpha)
        RenderSystem.enableBlend()
        RenderSystem.defaultBlendFunc()
        RenderSystem.blendFunc(GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA)
        blit(matrixStack, x, y, 0.0f, if (isChecked) 20.0f else 0.0f, 20, height, 32, 64)
        renderBg(matrixStack, minecraft, mouseX, mouseY)
        val i = 14737632
        Minecraft.getInstance().fontRenderer.drawString(
            matrixStack, message.string, x + 24.toFloat(), y + (height - 8) / 2f, 14737632 or MathHelper.ceil(
                alpha * 255.0f
            ) shl 24
        )
    }

    companion object {
        private val TEXTURE = ResourceLocation("textures/gui/checkbox.png")
    }
}